code/modules/reagents/reagent_containers/cups/drinks.dm

```diff
@@ -434,13 +434,92 @@
 	amount_per_transfer_from_this = 10
 	volume = 100
 	isGlass = FALSE
+	var/capped = TRUE
+	var/shaken = FALSE
+	var/cap_icon = 'icons/obj/drinks/bottles.dmi'
+	var/cap_icon_state = "shaker_cap"
+	var/mutable_appearance/cap_overlay
 
 /obj/item/reagent_containers/cup/glass/shaker/Initialize(mapload)
 	. = ..()
 	if(prob(10))
 		name = "\improper Nanotrasen 20th Anniversary Shaker"
 		desc += " It has an emblazoned Nanotrasen logo on it."
 		icon_state = "shaker_n"
+	cap_overlay = mutable_appearance(cap_icon, cap_icon_state)
+	if(capped)
+		spillable = FALSE
+		update_appearance()
+
+/obj/item/reagent_containers/cup/glass/shaker/update_overlays()
+	. = ..()
+	if(capped)
+		. += cap_overlay
+
+/obj/item/reagent_containers/cup/glass/shaker/examine(mob/user)
+	. = ..()
+	if(capped)
+		. += span_notice("The cap is firmly fitted and the shaker's ready to rumble. Alt-click to remove the cap.")
+	else
+		. += span_notice("The cap has been taken off and the shaker can be filled or emptied. Alt-click to put a cap on.")
+
+/obj/item/reagent_containers/cup/glass/shaker/AltClick(mob/user)
+	. = ..()
+	if(capped)
+		capped = FALSE
+		spillable = TRUE
+		animate(src, transform = null, time = 2, loop = 0)
+		balloon_alert(user, "cap removed...")
+		playsound(loc, 'sound/effects/can_open1.ogg', 50, TRUE)
+	else
+		capped = TRUE
+		spillable = FALSE
+		balloon_alert(user, "cap added...")
+		playsound(loc, 'sound/items/duct_tape_snap.ogg', 50, TRUE)
+	update_appearance()
+
+/obj/item/reagent_containers/cup/glass/shaker/is_refillable()
+	if(capped)
+		return FALSE
+	return ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/is_drainable()
+	if(capped)
+		return FALSE
+	return ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/attack(mob/target, mob/living/user, def_zone)
+	if(!target)
+		return
+
+	if(capped && reagents.total_volume && istype(target))
+		to_chat(user, span_warning("You must remove the cap before you can do that!"))
+		return
+
+	return ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/afterattack(obj/target, mob/living/user, proximity)
+	. |= AFTERATTACK_PROCESSED_ITEM
+
+	if(capped && (target.is_refillable() || target.is_drainable() || (reagents.total_volume && !user.combat_mode)))
+		to_chat(user, span_warning("You must remove the cap before you can do that!"))
+		return
+
+	return . | ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/attack_self(mob/user)
+	if(!is_drainable())
+		balloon_alert(user, "shaking...")
+		Shake(duration = 1 SECONDS)
+		if(reagents.total_volume)
+			playsound(loc, 'sound/effects/can_shake.ogg', 50, TRUE)
+		if(!do_after(user, 1 SECONDS, target = user))
+			return FALSE
+		shaken = TRUE
+		addtimer(VARSET_CALLBACK(src, shaken, FALSE), 2 SECONDS)
```

Comment:
```suggestion
		addtimer(VARSET_CALLBACK(src, shaken, FALSE), 2 SECONDS, TIMER_DELETE_ME)
```
Actually don't know if this is required but might stop something funny happening if you delete the shaker in the 2 seconds before the timer fires.